mitxgraders/helpers/math_array.py

```diff
@@ -0,0 +1,371 @@
+                    "to a {self.shape_name}".format(self=self))
+
+        elif isinstance(other, MathArray):
+            if self.shape == other.shape:
+                return super(MathArray, self).__add__(other)
+
+            msg = ("Cannot add/subtract a {self.description} with a {other.description}.").format(
+                self=self, other=other)
+            raise MathArrayError(msg)
+
+        raise TypeError("Cannot add/subtract object of {type} to a {self.description}.".format(
+            type=type(other), self=self))
+
+    def __radd__(self, other):
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return self.__add__(-1*other)
+
+    def __rsub__(self, other):
+        return (-self).__add__(other) # pylint: disable=invalid-unary-operand-type
+
+    def __mul__(self, other):
+        if isinstance(other, Number):
+            return super(MathArray, self).__mul__(other)
+
+        elif isinstance(other, IdentityMultiple):
+            return super(MathArray, self).__mul__(other.value)
+
+        elif isinstance(other, MathArray):
+            if self.ndim == 0:
+                return super(MathArray, other).__mul__(self.item())
+            elif other.ndim == 0:
+                return super(MathArray, self).__mul__(other.item())
+            elif self.ndim > 2 or other.ndim > 2:
+                raise MathArrayError("Multiplication of tensor arrays is not currently supported.")
+            try:
+                return np.dot(self, other)
+            except ValueError:
+                # vector-specific message mentions dot product
+                if self.ndim == 1 and other.ndim == 1:
+                    msg = ("Cannot calculate the dot product of a {self.description} "
+                           "with a {other.description}".format(self=self, other=other))
+                    raise MathArrayError(msg)
+                # general message:
+                msg = ("Cannot multiply a {self.description} with a {other.description}.").format(
+                    self=self, other=other)
+                raise MathArrayError(msg)
+
+        raise TypeError("Cannot multiply object of {type} with a {self.description}.".format(
+            type=type(other), self=self))
+
+    def __rmul__(self, other):
+        if isinstance(other, IdentityMultiple):
+            other = other.value
+        return super(MathArray, self).__rmul__(other)
+
+    def __pow__(self, other):
+        """
+        Matrix powers for MathArrays of dimension 0 and 2.
+        """
+        if self.ndim == 0:
+            if isinstance(other, Number):
+                return robust_pow(self.item(), other)
+            elif isinstance(other, MathArray):
+                return other.__rpow__(self.item())
+            else:
+                raise TypeError("Cannot raise a scalar to a {type} power".format(type=type(other)))
+
+        elif not self.ndim == 2:
```

Comment:
Can we add in a special case for `self.ndim == 1` and `other == 2` to take the norm of a vector?

Reply:
I think we should use `abs(v)` or `norm(v)` for that. (Which reminds me, I'm pretty sure I made `abs` a scalar-only function. I should fix that.)

(The only course I ever took where `v^2` meant `v dot v` for vectors was graduate stat mech ...)

Reply:
Really? `v^2 = v dot v` is really very common! If `abs` plays the same role though, that's fine by me.